tests/unit/test_sftp.py

```diff
@@ -0,0 +1,224 @@
+    """Test SFTP server without credentials."""
+    with pytest.raises(CredentialError, match="Missing credentials."):
+        SftpConnector(credentials=None)
+
+
+@pytest.mark.connect()
+def test_get_connection_without_rsa_key(mocker):
+    """Test SFTP `get_connection` method without ras_key."""
+    mock_transport = mocker.patch("viadot.sources.sftp.paramiko.Transport")
+    mock_sftp_client = mocker.patch(
+        "viadot.sources.sftp.paramiko.SFTPClient.from_transport"
+    )
+
+    connector = SftpConnector(credentials=variables["credentials"])
+    connector.get_connection()
+
+    mock_transport.assert_called_once_with((variables["credentials"]["hostname"], 999))
+    mock_transport().connect.assert_called_once_with(
+        None, variables["credentials"]["username"], variables["credentials"]["password"]
+    )
+    mock_sftp_client.assert_called_once()
+
+
+@pytest.mark.connect()
+def test_get_connection_with_rsa_key(mocker):
+    """Test SFTP `get_connection` method with ras_key."""
+    mock_ssh_client = mocker.patch("viadot.sources.sftp.paramiko.SSHClient")
+    mock_ssh_instance = mock_ssh_client.return_value
+    mock_ssh_connect = mocker.patch.object(mock_ssh_instance, "connect")
+    mock_rsa_key = mocker.patch(
+        "viadot.sources.sftp.paramiko.RSAKey.from_private_key",
+        return_value=mocker.Mock(),
+    )
+    mock_transport = mocker.Mock()
+
+    # Ensure the SSHClient's transport attribute isn't None
+    mock_ssh_instance._transport = mock_transport
+
+    dummy_rsa_key = (
+        "-----BEGIN RSA PRIVATE KEY-----\n###############...END RSA PRIVATE KEY-----"
+    )
+    keyfile = StringIO(dummy_rsa_key)
+    credentials = variables["credentials"]
+    credentials["rsa_key"] = keyfile.getvalue()
+
+    connector = SftpConnector(credentials=variables["credentials"])
+    connector.get_connection()
+
+    mock_rsa_key.assert_called_once()
+    mock_rsa_key.assert_called_once()
+    mock_ssh_connect.assert_called_once_with(
+        "", username="test_user", pkey=mock_rsa_key.return_value
+    )
+    assert connector.conn is not None
+
+
+@pytest.mark.functions()
+def test_to_df_with_csv(mocker):
+    """Test SFTP `to_df` method with csv."""
+    mock_getfo_file = mocker.patch.object(SftpConnector, "_getfo_file", autospec=True)
+    mock_getfo_file.return_value = BytesIO(b"col1,col2\n1,2\n3,4\n")
+
+    connector = SftpConnector(credentials=variables["credentials"])
+    df = connector.to_df(file_name="test.csv", sep=",")
+
+    assert isinstance(df, pd.DataFrame)
+    assert df.shape == (2, 4)
+    assert list(df.columns) == [
+        "col1",
+        "col2",
+        "_viadot_source",
+        "_viadot_downloaded_at_utc",
+    ]
+
+
+@pytest.mark.functions()
+def test_to_df_with_tsv(mocker):
+    """Test SFTP `to_df` method with tsv."""
+    mock_getfo_file = mocker.patch.object(SftpConnector, "_getfo_file", autospec=True)
+    mock_getfo_file.return_value = BytesIO(b"col1,col2\n1,2\n3,4\n")
```

Comment:
TSV means Tab Separated File; this is another CSV, so I don't think you're testing what you're saying you're testing here.

Reply:
because I have to cover several lines, TSV generation included.

Reply:
Test is removed